man/thumbnailer-admin.1
\#
\# Copyright (C) 2015 Canonical Ltd.
\#
\# This program is free software: you can redistribute it and/or modify
\# it under the terms of the GNU General Public License version 3 as
\# published by the Free Software Foundation.
\#
\# This program is distributed in the hope that it will be useful,
\# but WITHOUT ANY WARRANTY; without even the implied warranty of
\# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
\# GNU General Public License for more details.
\#
\# You should have received a copy of the GNU General Public License
\# along with this program.  If not, see <http://www.gnu.org/licenses/>.
\#
\# Authored by: Michi Henning <michi.henning@canonical.com>
\#

.TH thumbnailer\-admin 1 "05 August 2015" "Ubuntu" "Linux User's Manual"

.SH NAME
thumbnailer\-admin \- Thumbnailer administrative tool

.SH SYNOPSIS
.B thumbnailer\-admin
.I command
.RI [ options ]
.br

.SH DESCRIPTION
The \fBthumbnailer\-admin\fP program provides a simple administrative tool to access
the \fBthumbnailer\-service\fP(1), for example, to retrieve a thumbnail, display
statistics, or selectively clear one ore more of the service's internal caches.
.SS "Internal caches"
The thumbnailer service maintains three internal caches:
.TP
Image cache
The image cache stores images that are expensive to retrieve (namely
images downloaded from the image server (dash.ubuntu.com) or extracted from
video files). Images in this cache are kept at high resolution,
up to a limit of \fBmax\-thumbnail\-size\fP (usually 1920) pixels in the larger dimension (see
\fBthumbnailer\-settings\fR(5)). The image cache
is used to create thumbnails at smaller sizes that are then inserted
into the thumbnail cache (to avoid repeated download or extraction
of the original images).
.TP
Thumbnail cache
The thumbnail cache stores images that are scaled to a size that was previously requested
by a client. If a client asks for a thumbnail that was previously generated at the same size,
it is returned from this cache.
.TP
Failure cache
The failure cache contains the keys for media (local or remote) for which no thumbnail
could be generated, either because a local file does not contain artwork, or because
the image server has returned an authoritative answer that no image is available.
If a client asks for an image that is in the failure cache, the service does not
try to create a thumbnail for some period of time, to avoid expensive repeated
attempts to retrieve artwork that does not exist.
.P
Commands that can selectively be applied to these caches use the following \fIcache\-id\fP:
.TP
.B i
Image cache
.TP
.B t
Thumbnail cache
.TP
.B f
Failure cache
.SS "Sizes and scaling"
Thumbnails are never larger than \fBmax\-thumbnail\-size\fP in the larger dimension (usually 1920,
see \fBthumbnailer\-settings\fR(5)), even if a larger size is requested.

Commands that create thumbnails expect a size in pixels to be specified as follows:
.TP
\fIwidth\fP\fBx\fP\fIheight\fP
Create a thumbnail scaled to fit a bounding box of \fIwidth\fP and \fIheight\fP, for example,
\fB\-\-size 640x480\fP. If only a single dimension is provided, the thumbnail is scaled to fit into
a square bounding box of the specified size, for example, \fB\-s 256\fP.
.IP
Specifying a size with one dimension explicitly set to zero leaves that dimension unconstrained. For
example, \fB\-\-size 0x480\fP produces a thumbnail no taller than 480 pixels of any width,
(subject to the \fBmax\-thumbnail\-size\fP limit).
.IP
Specifying a size of zero (or not specifying any size) is equivalent to specifying \fBmax\-thumbnail\-size\fP
in both dimensions.
.IP
Scaling always maintains the aspect ratio of the original image, and ensures that a thumbnail
has a size of at least one pixel in the smaller dimension.
.IP
Requests for thumbnails that are larger than the original image are automatically reduced to the original
image size, that is, original artwork is never scaled up.
.SS "Media formats and output files"
Most common image formats, such as PNG, JPEG, BMP, and so on, are recognized. For streaming media, the
recognized formats depend on the installed gstreamer codecs.
.P
Thumbnails are always returned in PNG format, regardless of the encoding of the original image. The generated
thumbnail file is tagged with the requested (not actual) image size. For example,
.IP
.RS
.B thumbnailer\-admin get \-\-size=320x240 screen.jpg
.RE
.P
creates a file named \fBscreen_320x480.png\fP.

.SH COMMAND OVERVIEW
.TP
.B thumbnailer\-admin clear \fR[\fIcache\-id\fR]
Clear all internal caches.
.TP
.B thumbnailer\-admin stats \fR[\fIcache\-id\fR]
Display cache statistics.
.TP
.B thumbnailer\-admin zero\-stats \fR[\fIcache\-id\fR]
Zero statistics counters.
.TP
.B thumbnailer\-admin get file \fR[\fIdir\fR]
Get a thumbnail from a file.
.TP
.B thumbnailer\-admin get\-artist artist album \fR[\fIdir\fR]
Get an artist thumbnail from the image server.
.TP
.B thumbnailer\-admin get\-album artist album \fR[\fIdir\fR]
Get an album thumbnail from the image server.
.TP
.B thumbnailer\-admin compact
Compact internal caches.
.TP
.B thumbnailer\-admin shutdown
Shut down the thumbnailer service.

.SH COMMAND REFERENCE
.B thumbnailer\-admin \-\-help
.br
.B thumbnailer\-admin \-h
.RS
Display a general help message for all commands.
.RE

.P
.B thumbnailer\-admin clear \fR[\fIcache\-id\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.P
Clear all three internal caches. If \fIcache\-id\fP is provided, clear only the selected cache.
.RE

.P
.B thumbnailer\-admin stats \fR[\fIcache\-id\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.RS
.B \-\-verbose
.br
.B \-v
.RS
Show histogram of entry sizes.
.RE
.RE
.P
Display detailed cache statistics. If \fIcache\-id\fP is provided, limit the display to the selected cache.
.RE

.P
.B thumbnailer\-admin zero\-stats \fR[\fIcache\-id\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.P
Reset the statistics counters and timestamps to zero. If \fIcache\-id\fP is provided, reset the
statistics only for the selected cache.
.RE

.P
.B thumbnailer\-admin get file \fR[\fIdir\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.RS
.B \-\-size \fIpixels\fP
.br
.B \-\-s \fIpixels\fP
.RS
Set the thumbnail size.
.RE
.RE
.P
Extract a thumbnail from the specified media \fBfile\fP.
.P
If \fIdir\fP is specified, place the thumbnail into the specified directory (default: current directory).
.P
The output file name is \fIbasename\fB_\fIsize\fB.png\fR. For example, for a file \fBsong.mp3\fP, the output
file name is \fBsong_0x0.png\fP if no specific size was provided.
.RE

.P
.B thumbnailer\-admin get\-artist artist album \fR[\fIdir\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.RS
.B \-\-size \fIpixels\fP
.br
.B \-\-s \fIpixels\fP
.RS
Set the thumbnail size.
.RE
.RE
.P
Retrieve artist artwork for the specified \fBartist\fP and \fBalbum\fP. The album can be an empty string but,
because artist names alone can be ambiguous, specifying an album as well may yield a better result.
.P
If \fIdir\fP is specified, place the thumbnail into the specified directory (default: current directory).
.P
The output file name is \fIartist\fB_\fIalbum\fB_artist_\fIsize\fB.png\fR. For example, for Jethro Tull's
"Aqualung", the file name is \fBJethro Tull_Aqualung_artist_0x0.png\fP if no specific size was provided.
.RE

.P
.B thumbnailer\-admin get\-album artist album \fR[\fIdir\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.RS
.B \-\-size \fIpixels\fP
.br
.B \-\-s \fIpixels\fP
.RS
Set the thumbnail size.
.RE
.RE
.P
Retrieve album cover artwork for the specified \fBartist\fP and \fBalbum\fP. The artist can be an empty string but,
because album titles alone can be ambiguous, specifying an artist as well may yield a better result.
.P
If \fIdir\fP is specified, place the thumbnail into the specified directory (default: current directory).
.P
The output file name is \fIartist\fB_\fIalbum\fB_album_\fIsize\fB.png\fR. For example, for Jethro Tull's
"Aqualung", the file name is \fBJethro Tull_Aqualung_album_0x0.png\fP if no specific size was provided.
.RE

.P
.B thumbnailer\-admin compact \fR[\fIcache\-id\fR]
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.P
Compact all three internal caches so the take up the smallest possible amount of disk space. If \fIcache\-id\fP is provided, clear only the selected cache.
.P
After some time of use, the on\-disk size of a cache can be 25\(en30% larger than its nominal size. Compaction reclaims
the wasted space, but is slow and, while in progress, other thumbnail requests are delayed until compaction completes.
(It can take tens of seconds to compact a cache with a nominal size of 100 MB, depending on the speed of the hardware.)
.RE

.P
.B thumbnailer\-admin shutdown
.RS
Options:
.RS
.B \-\-help
.br
.B \-h
.RS
Show help message.
.RE
.RE
.P
Shut down the thumbnailer service. This is useful mainly after changing settings (which are read only once when the
service starts).
.RE

.SH "SEE ALSO"
.B thumbnailer\-service\fR(1)\fP, thumbnailer\-settings\fR(5)\fP


debian/changelog
thumbnailer (2.4+16.10.20160530.3-0ubuntu2) UNRELEASED; urgency=medium

  * REPLACE_ME: Placeholder entry after merging trunk.

 -- Michi Henning <michi.henning@canonical.com>  Wed, 01 Jun 2016 08:17:27 +1000

thumbnailer (2.4+16.10.20160530.3-0ubuntu1) yakkety; urgency=medium

  [ Michi Henning ]
  * Re-enabled copyright test on xenial and yakkety.

  [ CI Train Bot ]
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Mon, 30 May 2016 11:07:33 +0000

thumbnailer (2.4+16.04.20160321-0ubuntu2) yakkety; urgency=medium

  * No-change rebuild for boost soname change.

 -- Matthias Klose <doko@ubuntu.com>  Sat, 23 Apr 2016 18:50:26 +0000

thumbnailer (2.4+16.04.20160321-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * vs-thumb now uses a separate pipe instead of writing to stdout (LP: #1556835).

  [ CI Train Bot ]
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Mon, 21 Mar 2016 02:12:07 +0000

thumbnailer (2.4+16.04.20160302.1-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * Thumbnails now preserve transparency of source image (LP: #1532719, LP: #1541836).
  * Fixed incorrect tests when checking full-size cache hits (LP: #1540753).
  * Minor improvements to man pages.

  [ CI Train Bot ]
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Wed, 02 Mar 2016 05:20:31 +0000

thumbnailer (2.4+16.04.20160226-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * Fixed incorrect build dependency for settings-defaults.h.
  * Re-enabled tests on powerpc and s390x, except for qml test.
  * Disabled vs-thumb test on powerpc64le due to gstreamer codec bug.
  * Added version check on start-up and automatic cache wipe if cache
    encoding version changes.
  * Changed local audio extraction for better performance: 12 times
    faster on Mako, 5 times faster on desktop.
  * thumbnailer-admin now works with relative paths (LP: #1538774).

  [ James Henstridge ]
  * Fixed thumbnail extraction from large animated GIF files (LP: #1527315).

  [ CI Train Bot ]
  * No-change rebuild.

 -- James Henstridge <james.henstridge@canonical.com>  Fri, 26 Feb 2016 03:27:56 +0000

thumbnailer (2.3+16.04.20160210-0ubuntu1) xenial; urgency=medium

  [ James Henstridge ]
  * When a thumbnail can not be loaded, return an error from the image
    provider rather than substituting in a fallback image. Users can
    detect the problem via the status property of the QML Image
    component and provide their own fallback handling. (LP: #1324142)

 -- Pawel Stolowski <ci-train-bot@canonical.com>  Wed, 10 Feb 2016 00:51:48 +0000

thumbnailer (2.3+16.04.20160108.2-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * Fixed incorrect display of cache policy by "thumbnailer-admin stats".
  * Limit GIF file size to 2 MB (LP: #1527315).
  * Fix non-working config settings for client-side API (LP: #1528058).

  [ CI Train Bot ]
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Fri, 08 Jan 2016 12:43:17 +0000

thumbnailer (2.3+16.04.20160107.1-0ubuntu1) xenial; urgency=medium

  * Fix for lp:1531038. Much better user experience this way. (LP:
    #1531038)

 -- Michi Henning <michi.henning@canonical.com>  Thu, 07 Jan 2016 06:33:30 +0000

thumbnailer (2.3+16.04.20160107-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * Ignored failing tests on PPC due to gstreamer problems.
  * Fixed crash that happened if the service was shut down
    while there were requests still processing in the
    thread pools.
  * Fixed incorrect handling of 400 (Bad Request) response from
    remote server (LP: #1512608).
  * Passing an invalid size to the QML or Qt APIs is now an error.
  * Added new gsettings key "log-level". Amount of noise in the logs
    is now reduced. Set log-level to 2 for verbose logging.

  [ CI Train Bot ]
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Thu, 07 Jan 2016 04:36:39 +0000

thumbnailer (2.3+16.04.20151102.2-0ubuntu1) xenial; urgency=medium

  [ Michi Henning ]
  * "thumbnailer-admin clear f" now also resets the network retry timeout.
  * API documentation and man page improvements.
  * Videos with embedded artwork now display the artwork thumbnail instead
    of an extracted still frame (LP: #1510801).
  * Added cancelled() accessor to Request class.
  * Requests made via the Qt and QML APIs are correctly rate-limited now.
  * Added trace-client key to gsettings.
  * Service log now shows statistics on start-up.
  * Resetting the stats now also resets the network failure retry timout.
  * Fix for cover art extraction if only a preview cover image is available.

  [ Michael Sheldon ]
  * Added more robust content type detection (LP: #1499341).

  [ CI Train Bot ]
  * debian/libthumbnailer-qt1.0.symbols: update to released version.
  * No-change rebuild.

 -- Michi Henning <michi.henning@canonical.com>  Mon, 02 Nov 2015 12:36:56 +0000

thumbnailer (2.3+15.10.20150915.1-0ubuntu1) wily; urgency=medium

  [ Michi Henning ]
  * Added man pages for thumbnailer-admin, thumbnailer-service,
    and thumbnailer-settings.
  * Removed a bunch of lintian warnings and errors.
  * Removed source for persistent cache.
  * Added inactivity timeout to admin interface, so the service
    exits after 30 seconds if it is started only to, for example,
    retrieve the stats. Activity on the admin interface now keeps
    the service alive the same way activity on the thumbnailer interface does.
  * Added average hit and miss run length to stats returned by
    thumbnailer-admin.
  * Changed semantics for requests with (0,0) size. All requests are now
    clamped to a 1920x1920 bounding box.
  * Fixed incorrect timeout semantics for unexpected network errors.

  [ James Henstridge ]
  * Renamed QML package to qml-module-XXX format and add transitional
    package. (LP: #1342031)
  * Port QML module to use libthumbnailer-qt.

  [ Xavi Garcia ]

src/thumbnailer-admin/thumbnailer-admin.cpp
/*
 * Copyright (C) 2015 Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authored by: Michi Henning <michi.henning@canonical.com>
 */

#include "clear.h"
#include "dbus_connection.h"
#include "get_local_thumbnail.h"
#include "get_remote_thumbnail.h"
#include "show_stats.h"
#include "shutdown.h"

#include <boost/filesystem.hpp>

#include <functional>
#include <iomanip>
#include <iostream>
#include <sstream>

#include <stdio.h>

using namespace std;
using namespace unity::thumbnailer::tools;

namespace
{

string prog_name;

template<typename A>
Action::UPtr create_action(QCommandLineParser& parser)
{
    return Action::UPtr(new A(parser));
}

typedef map<char const*, pair<Action::UPtr(*)(QCommandLineParser&), char const*>> ActionMap;

// Table that maps commands to their actions.
// Add new commands to this table, and implement each command in
// a class derived from Action.

ActionMap const valid_actions =
{
    { "stats",       { &create_action<ShowStats>,          "Show statistics" } },
    { "zero-stats",  { &create_action<Clear>,              "Zero statistics counters" } },
    { "get",         { &create_action<GetLocalThumbnail>,  "Get thumbnail from local file" } },
    { "get-artist",  { &create_action<GetRemoteThumbnail>, "Get artist thumbnail" } },
    { "get-album",   { &create_action<GetRemoteThumbnail>, "Get album thumbnail" } },
    { "clear",       { &create_action<Clear>,              "Clear caches" } },
    { "compact",     { &create_action<Clear>,              "Compact caches" } },
    { "shutdown",    { &create_action<Shutdown>,           "Shut down thumbnailer service" } }
};

QString command_summary()
{
    stringstream s;
    s << "Commands:" << endl;
    for (auto const& a : valid_actions)
    {
        s << "  " << setw(11) << left << a.first << " " << a.second.second << endl;
    }
    return QString::fromStdString(s.str());
}

// Check if we have a valid command. If so, instantiate the
// corresponding action and execute it.

void parse_and_execute()
{
    QCommandLineParser parser;
    parser.setApplicationDescription(QStringLiteral("Thumbnailer administrative tool"));
    parser.addPositionalArgument(QStringLiteral("command"), QStringLiteral("The command to execute."));
    parser.parse(QCoreApplication::arguments());

    auto args = parser.positionalArguments();
    if (args.empty())
    {
        throw parser.helpText() + "\n" + command_summary();
    }
    QString cmd = args.first();
    for (auto a : valid_actions)
    {
        if (QString(a.first) == cmd)
        {
            auto action = a.second.first(parser);
            DBusConnection conn;
            action->run(conn);
            return;
        }
    }
    fprintf(stderr, "%s: %s: invalid command\n", prog_name.c_str(), qPrintable(cmd));
    throw parser.helpText() + "\n" + command_summary();
}

}  // namespace

int main(int argc, char* argv[])
{
    int rc = EXIT_FAILURE;
    try
    {
        QCoreApplication app(argc, argv);
        boost::filesystem::path p = app.applicationName().toStdString();
        prog_name = p.filename().native();
        parse_and_execute();
        rc = EXIT_SUCCESS;
    }
    // LCOV_EXCL_START
    catch (std::exception const& e)
    {
        cerr << prog_name << ": " << e.what() << endl;
    }
    catch (QString const& msg)
    {
        cerr << prog_name << ": " << msg.toStdString() << endl;
    }
    catch (string const& msg)
    {
        cerr << prog_name << ": " << msg << endl;
    }
    catch (char const* msg)
    {
        cerr << prog_name << ": " << msg << endl;
    }
    // LCOV_EXCL_STOP
    // No catch for ... here. It's better to dump core.
    return rc;
}
